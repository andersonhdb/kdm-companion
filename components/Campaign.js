import React from 'react'
import { View } from '@shoutem/ui'
import { observer, inject } from 'mobx-react/native'
import { values } from 'mobx'
import CheckboxListItem from './common/CheckboxListItem'
import LockedCheckboxListItem from './common/LockedCheckboxListItem'

import colors from '../src/colors'

@inject(({ store }) => ({
  campaign: store.selectedCampaign,
  subscription: store.subscription,
}))
@observer
export default class Expansions extends React.Component {
  constructor(props) {
    super(props)
  }

  selectCampaign = id => {
    if (id === 'potl' || this.props.subscription.hasActiveSubscription()) {
      this.props.campaign.setCampaignType(id)
    } else {
      this.props.subscribe()
    }
  }

  render() {
    let Checkbox
    if (this.props.subscription.hasActiveSubscription()) {
      Checkbox = CheckboxListItem
    } else {
      Checkbox = LockedCheckboxListItem
    }
    return (
      <View>
        <CheckboxListItem
          styleName="title"
          onPressItem={() => this.selectCampaign('potl')}
          title="People of the Lantern (Core)"
          id="core"
          selected={this.props.campaign.type.id === 'potl'}
        />
        <Checkbox
          styleName="title"
          onPressItem={() => this.selectCampaign('pots')}
          title="People of the Stars (Dragon King)"
          id="pots"
          selected={this.props.campaign.type.id === 'pots'}
        />
        <Checkbox
          styleName="title"
          onPressItem={() => this.selectCampaign('potsun')}
          title="People of the Sun (Sunstalker)"
          id="potsun"
          selected={this.props.campaign.type.id === 'potsun'}
        />
      </View>
    )
  }
}

const styles = {}
