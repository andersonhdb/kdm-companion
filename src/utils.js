import R from 'ramda'
import { values } from 'mobx'

export const capitalize = str => str[0].toUpperCase() + str.slice(1)

export function uuid() {
  return ([1e7] + -1e3 + -4e3 + -8e3 + -1e11).replace(
    /[018]/g,
    a => (a ^ ((Math.random() * 16) >> (a / 4))).toString(16)[0]
  )
}

export function expansionFilter(map, expansion) {
  return R.filter(item => item.expansion.id === expansion.id, values(map))
}

/**
 *
 * @param {object} base object to merge data into (eg: getSnapshot(SettlementStat.create()))
 * @param {array} values array of objects with base attributes to merge (eg: values(self.innovations))
 * @param {array} attribute attribute from each values item to pick to merge into the base (eg: ['settlement', 'newborn'] )
 */
export function objectMerge(base, values, attribute) {
  let keys = Object.keys(base)
  let result = values.reduce((merged, item) => {
    if (R.path(attribute, item)) {
      // combine the 2 objects: current accumulated base and the one from the values item.
      return keys.reduce((keysAcc, key) => {
        return {
          ...keysAcc,
          [key]:
            typeof merged[key] === 'string'
              ? R.path(attribute, item)[key].length > 0
                ? merged[key] + R.path(attribute, item)[key] + '\n'
                : merged[key]
              : merged[key] + R.path(attribute, item)[key],
        }
      }, {})
    } else {
      return merged
    }
  }, base)

  return result
}
