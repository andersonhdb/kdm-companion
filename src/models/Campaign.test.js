import { getSnapshot } from 'mobx-state-tree'
import { keys, values, observable } from 'mobx'
import { Campaign } from './Campaign'
import RootStore from './RootStore'
import * as utils from '../utils'

afterEach(() => {
  jest.restoreAllMocks()
})

describe('.create', () => {
  it('creates a barebones campaign', () => {
    const uuid = '960aa7ae-598a-49eb-888b-c6baa0006751'
    let uuidMock = jest.spyOn(utils, 'uuid')
    uuidMock.mockImplementation(() => uuid)

    let campaign = Campaign.create()

    expect(campaign).toBeDefined()
    expect(campaign.id).toBe(uuid)
    expect(getSnapshot(campaign)).toMatchSnapshot()
  })
})

describe('with RootStore', () => {
  let store
  beforeEach(() => {
    const uuid = '960aa7ae-598a-49eb-888b-c6baa0006751'
    let uuidMock = jest.spyOn(utils, 'uuid')
    uuidMock.mockImplementation(() => uuid)
    store = RootStore.create()
  })

  describe('.selectLocation', () => {
    it('adds location without endeavors', () => {
      let location = { id: 'catarium' }
      store.selectedCampaign.selectLocation(location)

      expect(getSnapshot(store.selectedCampaign)).toMatchSnapshot()
    })
    it('removes location without endeavors', () => {
      let location = { id: 'catarium' }
      store.selectedCampaign.selectLocation(location)
      store.selectedCampaign.selectLocation(location) // this removes the location.

      expect(getSnapshot(store.selectedCampaign)).toMatchSnapshot()
    })

    it('adds location with endeavors', () => {
      let location = { id: 'lantern_hoard' }
      store.selectedCampaign.selectLocation(location)

      expect(store.selectedCampaign.endeavors).toHaveLength(5)
      expect(store.selectedCampaign.endeavors).toMatchObject([
        { id: 'innovate' },
        { id: 'shared_experience' },
        { id: 'build_bone_smith' },
        { id: 'build_skinnery' },
        { id: 'build_organ_grinder' },
      ])
    })
    it('removes location with endeavors', () => {
      let location = { id: 'stone_circle' }
      store.selectedCampaign.selectLocation(location)
      store.selectedCampaign.selectLocation(location) // this removes the location.

      expect(store.selectedCampaign.endeavors).toHaveLength(0)
    })
    it('can add locations with multiple endeavors', () => {
      let location = { id: 'weapon_crafter' }
      store.selectedCampaign.selectLocation(location)

      expect(store.selectedCampaign.endeavors).toHaveLength(2)
    })
  })

  describe('.selectInnovation', () => {
    it('can add an innovation', () => {
      let innovation = { id: 'inner_lantern' }
      store.selectedCampaign.selectInnovation(innovation)

      expect(store.selectedCampaign.innovations.size).toEqual(1)
      expect(
        store.selectedCampaign.innovations.has('inner_lantern')
      ).toBeTruthy()
    })
    it('can remove an innovation', () => {
      let innovation = { id: 'inner_lantern' }
      store.selectedCampaign.selectInnovation(innovation)
      store.selectedCampaign.selectInnovation(innovation) // this removes the innovation

      expect(store.selectedCampaign.innovations.size).toEqual(0)
    })

    it('adds innovation with bonuses', () => {
      let innovation = { id: 'cooking' }
      store.selectedCampaign.selectInnovation(innovation)

      expect(store.selectedCampaign.bonuses).toHaveLength(1)

      expect(store.selectedCampaign.bonuses[0]).toMatchObject({
        description: 'At the start of the Settlement phase gain +1 endeavor',
      })
      expect(getSnapshot(store.selectedCampaign)).toMatchSnapshot()
    })
    it('removes innovation with bonuses', () => {
      let innovation = { id: 'cooking' }
      store.selectedCampaign.selectInnovation(innovation)
      store.selectedCampaign.selectInnovation(innovation) // this removes the innovation

      expect(store.selectedCampaign.bonuses).toHaveLength(0)

      expect(getSnapshot(store.selectedCampaign)).toMatchSnapshot()
    })

    it('handles innovation with endeavors', () => {
      let innovation = { id: 'partnership' }
      store.selectedCampaign.selectInnovation(innovation)

      expect(store.selectedCampaign.endeavors).toHaveLength(1)
      expect(store.selectedCampaign.endeavors).toMatchObject([
        { id: 'partnership' },
      ])
    })

    it('handles removing innovation with endeavors', () => {
      let innovation = { id: 'partnership' }
      store.selectedCampaign.selectInnovation(innovation)
      store.selectedCampaign.selectInnovation(innovation)

      expect(store.selectedCampaign.endeavors).toHaveLength(0)
    })
  })

  describe('.selectHunt', () => {
    it('sets the hunt', () => {
      store.selectedCampaign.selectHunt({
        monster_id: 'white_lion',
        level_id: '1',
      })

      expect(store.selectedCampaign.hunting.monster).toHaveProperty(
        'id',
        'white_lion'
      )
      expect(store.selectedCampaign.hunting).toHaveProperty('level', '1')
    })

    it('sets the showdown', () => {
      store.selectedCampaign.selectHunt({
        monster_id: 'white_lion',
        level_id: '1',
      })

      expect(store.selectedCampaign.showdown.monster).toHaveProperty(
        'id',
        'white_lion'
      )
      expect(store.selectedCampaign.showdown).toHaveProperty('level', '1')
    })
  })

  describe('.selectShowdown', () => {
    it('sets the showdown', () => {
      store.selectedCampaign.selectShowdown({
        monster_id: 'white_lion',
        level_id: '2',
      })

      expect(store.selectedCampaign.showdown.monster).toHaveProperty(
        'id',
        'white_lion'
      )
      expect(store.selectedCampaign.showdown).toHaveProperty('level', '2')
    })
  })

  describe('.selectExpansion', () => {
    it('does not remove core expansion', () => {
      store.selectedCampaign.selectExpansion({ id: 'core' })

      expect(store.selectedCampaign.expansions.get('core')).toBeDefined()
    })

    it('adds expansion', () => {
      store.selectedCampaign.selectExpansion({ id: 'dk' })

      expect(store.selectedCampaign.expansions.get('dk')).toBeDefined()
    })

    it('removes expansion', () => {
      store.selectedCampaign.selectExpansion({ id: 'dk' })
      store.selectedCampaign.selectExpansion({ id: 'dk' })

      expect(store.selectedCampaign.expansions.get('dk')).toBeUndefined()
    })

    it('resets expansion related data when removing expansion', () => {
      let campaign = { id: 'dk' }
      store.selectedCampaign.selectExpansion(campaign)
      store.selectedCampaign.selectLocation({ id: 'throne' })
      store.selectedCampaign.selectInnovation({ id: 'radiating orb' })
      store.selectedCampaign.setResourceCount({ id: 'hardened ribs' }, 2) // dk monster resource
      store.selectedCampaign.setResourceCount({ id: 'pituitary gland' }, 2) // dk expansion strange resource
      store.selectedCampaign.selectHunt({ monster_id: 'dk', level_id: '1' })

      // remove the expansion
      store.selectedCampaign.selectExpansion(campaign)

      expect(store.selectedCampaign.expansions.get('dk')).toBeUndefined()
      expect(store.selectedCampaign.hunting).toBeNull()
      expect(store.selectedCampaign.locations.get('throne')).toBeUndefined()
      expect(
        store.selectedCampaign.innovations.get('radiating orb')
      ).toBeUndefined()
      expect(
        store.selectedCampaign.stored_resources.get('pituitary gland')
      ).toHaveProperty('quantity', 0)
      expect(
        store.selectedCampaign.stored_resources.get('hardened ribs')
      ).toHaveProperty('quantity', 0)

      expect(getSnapshot(store.selectedCampaign)).toMatchSnapshot()
    })
  })

  describe('.selectPrinciple', () => {
    it('adds a principle', () => {
      store.selectedCampaign.selectPrinciple('death', { id: 'cannibalize' })

      expect(getSnapshot(store.selectedCampaign)).toMatchSnapshot()
    })

    it("accounts for principle's bonus", () => {
      store.selectedCampaign.selectPrinciple('death', { id: 'cannibalize' })

      expect(store.selectedCampaign.bonuses).toHaveLength(1)
      expect(store.selectedCampaign.bonuses[0]).toMatchObject({
        description: 'Whenever a survivor dies, gain 1 random basic resource',
      })
    })

    it("removes a principle with it's bonus", () => {
      store.selectedCampaign.selectPrinciple('death', { id: 'cannibalize' })
      store.selectedCampaign.selectPrinciple('death', { id: null })

      expect(store.selectedCampaign.bonuses).toHaveLength(0)
      expect(getSnapshot(store.selectedCampaign)).toMatchSnapshot()
    })
  })

  describe('.setResourceCount', () => {
    it('creates new stored resource', () => {
      expect(
        store.selectedCampaign.stored_resources.get('sinew')
      ).toBeUndefined()

      store.selectedCampaign.setResourceCount({ id: 'sinew' }, 2)

      expect(store.selectedCampaign.stored_resources.get('sinew')).toBeDefined()
      expect(
        store.selectedCampaign.stored_resources.get('sinew')
      ).toHaveProperty('quantity', 2)
    })

    it('updates existing resource', () => {
      store.selectedCampaign.setResourceCount({ id: 'sinew' }, 2)
      store.selectedCampaign.setResourceCount({ id: 'sinew' }, 4)

      expect(
        store.selectedCampaign.stored_resources.get('sinew')
      ).toHaveProperty('quantity', 4)
    })
  })

  describe('.huntingMonsterLevel', () => {
    it('gets the monster level', () => {
      store.selectedCampaign.selectHunt({
        monster_id: 'white_lion',
        level_id: '3',
      })

      expect(
        getSnapshot(store.selectedCampaign.huntingMonsterLevel)
      ).toMatchSnapshot()
    })
  })

  describe('.showdownMonsterLevel', () => {
    it('gets the monster level', () => {
      store.selectedCampaign.selectHunt({
        monster_id: 'white_lion',
        level_id: '3',
      })

      expect(
        getSnapshot(store.selectedCampaign.showdownMonsterLevel)
      ).toMatchSnapshot()
    })
  })

  describe('.selectedExpansionFilter', () => {
    it('filters regular objects', () => {
      store.selectedCampaign.selectExpansion({ id: 'dk' })

      let items = [
        { id: 'test1', expansion: { id: 'core' } },
        { id: 'test2', expansion: { id: 'dk' } },
        { id: 'test3', expansion: { id: 'gorm' } },
      ]

      let filtered = store.selectedCampaign.selectedExpansionFilter(items)
      expect(filtered).toEqual([
        { id: 'test1', expansion: { id: 'core' } },
        { id: 'test2', expansion: { id: 'dk' } },
      ])
    })

    it('filters maps', () => {
      store.selectedCampaign.selectExpansion({ id: 'dk' })

      let items = observable.map(
        new Map([
          ['test1', { id: 'test1', expansion: { id: 'core' } }],
          ['test2', { id: 'test2', expansion: { id: 'dk' } }],
          ['test3', { id: 'test3', expansion: { id: 'gorm' } }],
        ])
      )

      let filtered = store.selectedCampaign.selectedExpansionFilter(items)
      expect(filtered).toEqual([
        { id: 'test1', expansion: { id: 'core' } },
        { id: 'test2', expansion: { id: 'dk' } },
      ])
    })

    it('filters campaigns', () => {
      store.selectedCampaign.setCampaignType('pots')

      let items = observable.map(
        new Map([
          ['test0', { id: 'test0' }],
          ['test1', { id: 'test1', expansion: { id: 'core' } }],
          ['test2', { id: 'test2', expansion: { id: 'dk' } }],
          ['test3', { id: 'test3', expansion: { id: 'gorm' } }],
          [
            'test4',
            {
              id: 'test4',
              expansion: { id: 'core' },
              campaign: { id: 'potl' },
            },
          ],
          [
            'test5',
            { id: 'test5', expansion: { id: 'dk' }, campaign: { id: 'pots' } },
          ],
        ])
      )

      let filtered = store.selectedCampaign.selectedExpansionFilter(items)
      expect(filtered).toEqual([
        { id: 'test0' },
        { id: 'test1', expansion: { id: 'core' } },
        { id: 'test2', expansion: { id: 'dk' } },
        { id: 'test5', expansion: { id: 'dk' }, campaign: { id: 'pots' } },
      ])

      store.selectedCampaign.setCampaignType('potl')
      filtered = store.selectedCampaign.selectedExpansionFilter(items)
      expect(filtered).toEqual([
        { id: 'test0' },
        { id: 'test1', expansion: { id: 'core' } },
        { id: 'test2', expansion: { id: 'dk' } },
        { id: 'test4', expansion: { id: 'core' }, campaign: { id: 'potl' } },
      ])
    })
  })

  describe('.expansionContent', () => {
    it('gets all content', () => {
      store.selectedCampaign.selectExpansion({ id: 'dk' })
      store.selectedCampaign.selectLocation({ id: 'throne' })
      store.selectedCampaign.selectInnovation({ id: 'radiating orb' })
      store.selectedCampaign.setResourceCount({ id: 'hardened ribs' }, 2) // dk monster resource
      store.selectedCampaign.setResourceCount({ id: 'pituitary gland' }, 2) // dk expansion strange resource

      expect(
        store.selectedCampaign.expansionContent({ id: 'dk' })
      ).toMatchSnapshot()
    })
  })

  describe('.hasSOTF', () => {
    it('has not sotf', () => {
      expect(store.selectedCampaign.hasSOTF).toBeFalsy()
    })

    it('has sotf', () => {
      store.selectedCampaign.selectPrinciple('newlife', { id: 'sotf' })
      expect(store.selectedCampaign.hasSOTF).toBeTruthy()
    })
  })

  describe('.endeavors', () => {
    it('filters endeavors that match the not_location recipe', () => {
      let innovation = { id: 'scrap_smelting' }
      store.selectedCampaign.selectInnovation(innovation)

      expect(store.selectedCampaign.endeavors).toHaveLength(2)

      let location = { id: 'blacksmith' }
      store.selectedCampaign.selectLocation(location)

      expect(store.selectedCampaign.endeavors).toHaveLength(1)
      expect(store.selectedCampaign.endeavors).toMatchObject([
        { id: 'purification' },
      ])
    })

    it('filters endeavors that do not match the innovation recipe', () => {
      let location = { id: 'leather_worker' }
      store.selectedCampaign.selectLocation(location)

      expect(store.selectedCampaign.endeavors).toHaveLength(0) // leather_making requires ammonia

      let innovation = { id: 'ammonia' }
      store.selectedCampaign.selectInnovation(innovation)

      expect(store.selectedCampaign.endeavors).toHaveLength(1)
      expect(store.selectedCampaign.endeavors).toMatchObject([
        { id: 'leather_making' },
      ])
    })

    it('filters endeavors that match the not_innovation recipe', () => {
      let location = { id: 'weapon_crafter' }
      store.selectedCampaign.selectLocation(location)

      expect(store.selectedCampaign.endeavors).toHaveLength(2)

      let innovation = { id: 'scrap_smelting' }
      store.selectedCampaign.selectInnovation(innovation)

      expect(store.selectedCampaign.endeavors).toHaveLength(3) // 1 from weapon_crafter, 2 from scrap_smelting
      expect(store.selectedCampaign.endeavors).toMatchObject([
        { id: 'purification' },
        { id: 'build_blacksmith' },
        { id: 'scrap_scavenge' },
      ])
    })
  })
})
