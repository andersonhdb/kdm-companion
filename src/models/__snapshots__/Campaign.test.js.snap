// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.create creates a barebones campaign 1`] = `
Object {
  "bonuses": Object {},
  "endeavors": Object {},
  "expansions": Object {
    "core": "core",
  },
  "hunting": null,
  "id": "new",
  "innovations": Object {},
  "locations": Object {},
  "notes": "",
  "principles": Object {
    "conviction": null,
    "death": null,
    "newlife": null,
    "society": null,
  },
  "settlement": Object {
    "departing": null,
    "name": "New Settlement",
    "newborn": null,
    "showdown": null,
    "survivalLimit": 1,
    "survivors": Object {},
  },
  "showdown": null,
  "stored_resources": Object {},
}
`;

exports[`with RootStore .expansionContent gets all content 1`] = `
Object {
  "innovations": Array [
    Object {
      "endeavors": Array [],
      "expansion": "dk",
      "id": "radiating orb",
      "keywords": Array [
        "science",
      ],
      "name": "Radiating Orb",
      "providesBonuses": Array [],
      "settlement": Object {
        "departing": Object {
          "accuracy": null,
          "courage": null,
          "description": Array [
            "**Radiating Orb**: Survivors with a constellation gain +1 survival",
          ],
          "evasion": null,
          "hunt xp": null,
          "insanity": null,
          "strength": null,
          "survival": 1,
          "understanding": null,
        },
        "name": null,
        "newborn": Object {
          "accuracy": null,
          "courage": null,
          "description": null,
          "evasion": null,
          "hunt xp": null,
          "insanity": null,
          "strength": null,
          "survival": 1,
          "understanding": null,
        },
        "showdown": null,
        "survivalLimit": null,
        "survivors": Object {},
      },
    },
  ],
  "length": 4,
  "locations": Array [
    Object {
      "endeavors": Array [
        Object {
          "expansion": "dk",
          "id": "innovate",
          "name": "Innovate",
          "recipe": Object {
            "innovation": null,
            "items": Array [
              Object {
                "gear": null,
                "keyword": null,
                "quantity": 1,
                "resource": "endeavor",
              },
              Object {
                "gear": null,
                "keyword": "bone",
                "quantity": 1,
                "resource": null,
              },
              Object {
                "gear": null,
                "keyword": "organ",
                "quantity": 1,
                "resource": null,
              },
              Object {
                "gear": null,
                "keyword": "hide",
                "quantity": 1,
                "resource": null,
              },
            ],
            "location": "throne",
            "not_innovation": null,
            "not_location": null,
          },
        },
        Object {
          "expansion": "dk",
          "id": "fear and trembling",
          "name": "Fear and Trembling",
          "recipe": Object {
            "innovation": null,
            "items": Array [
              Object {
                "gear": null,
                "keyword": null,
                "quantity": 1,
                "resource": "endeavor",
              },
            ],
            "location": "throne",
            "not_innovation": null,
            "not_location": null,
          },
        },
        Object {
          "expansion": "dk",
          "id": "build_bone_smith",
          "name": "Build - Bone Smith",
          "recipe": Object {
            "innovation": null,
            "items": Array [
              Object {
                "gear": null,
                "keyword": null,
                "quantity": 1,
                "resource": "endeavor",
              },
            ],
            "location": "throne",
            "not_innovation": null,
            "not_location": "bone_smith",
          },
        },
        Object {
          "expansion": "dk",
          "id": "build_skinnery",
          "name": "Build - Skinnery",
          "recipe": Object {
            "innovation": null,
            "items": Array [
              Object {
                "gear": null,
                "keyword": null,
                "quantity": 1,
                "resource": "endeavor",
              },
            ],
            "location": "throne",
            "not_innovation": null,
            "not_location": "skinnery",
          },
        },
        Object {
          "expansion": "dk",
          "id": "build_organ_grinder",
          "name": "Build - Organ Grinder",
          "recipe": Object {
            "innovation": null,
            "items": Array [
              Object {
                "gear": null,
                "keyword": null,
                "quantity": 1,
                "resource": "endeavor",
              },
            ],
            "location": "throne",
            "not_innovation": null,
            "not_location": "organ_grinder",
          },
        },
      ],
      "expansion": "dk",
      "id": "throne",
      "name": "Throne",
    },
  ],
  "resources": Array [
    Object {
      "id": "hardened ribs",
      "quantity": 2,
      "resource": "hardened ribs",
    },
    Object {
      "id": "pituitary gland",
      "quantity": 2,
      "resource": "pituitary gland",
    },
  ],
}
`;

exports[`with RootStore .huntingMonsterLevel gets the monster level 1`] = `
Object {
  "accuracy": 2,
  "damage": 2,
  "deck": Object {
    "A": 9,
    "B": 10,
    "L": 2,
    "S": Array [
      "Cunning",
      "Merciless",
    ],
  },
  "evasion": 0,
  "huntExtra": null,
  "huntboard": "MMBBMOBMMBBXS",
  "id": "3",
  "luck": 1,
  "movement": 8,
  "name": "3",
  "speed": 2,
  "toughness": 14,
  "type": null,
}
`;

exports[`with RootStore .selectExpansion resets expansion related data when removing expansion 1`] = `
Object {
  "bonuses": Object {},
  "endeavors": Object {},
  "expansions": Object {
    "core": "core",
  },
  "hunting": null,
  "id": "new",
  "innovations": Object {},
  "locations": Object {},
  "notes": "",
  "principles": Object {
    "conviction": null,
    "death": null,
    "newlife": null,
    "society": null,
  },
  "settlement": Object {
    "departing": Object {
      "accuracy": 0,
      "courage": 0,
      "description": Array [],
      "evasion": 0,
      "hunt xp": 0,
      "insanity": 0,
      "strength": 0,
      "survival": 0,
      "understanding": 0,
    },
    "name": "New Settlement",
    "newborn": Object {
      "accuracy": 0,
      "courage": 0,
      "description": Array [],
      "evasion": 0,
      "hunt xp": 0,
      "insanity": 0,
      "strength": 0,
      "survival": 0,
      "understanding": 0,
    },
    "showdown": null,
    "survivalLimit": 1,
    "survivors": Object {},
  },
  "showdown": null,
  "stored_resources": Object {
    "hardened ribs": Object {
      "id": "hardened ribs",
      "quantity": 0,
      "resource": "hardened ribs",
    },
    "pituitary gland": Object {
      "id": "pituitary gland",
      "quantity": 0,
      "resource": "pituitary gland",
    },
  },
}
`;

exports[`with RootStore .selectInnovation adds innovation with bonuses 1`] = `
Object {
  "bonuses": Object {
    "cooking": Object {
      "description": Array [
        "At the start of the Settlement phase gain +1 endeavor",
      ],
      "id": "cooking",
      "name": "Cooking",
    },
  },
  "endeavors": Object {
    "cooking": Object {
      "expansion": "core",
      "id": "cooking",
      "name": "Cooking",
      "recipe": Object {
        "innovation": "cooking",
        "items": Array [
          Object {
            "gear": null,
            "keyword": null,
            "quantity": 1,
            "resource": "endeavor",
          },
        ],
        "location": null,
        "not_innovation": null,
        "not_location": null,
      },
    },
    "cooking_meal": Object {
      "expansion": "core",
      "id": "cooking_meal",
      "name": "Cook stone nose gruel",
      "recipe": Object {
        "innovation": "cooking",
        "items": Array [
          Object {
            "gear": null,
            "keyword": null,
            "quantity": 1,
            "resource": "endeavor",
          },
          Object {
            "gear": null,
            "keyword": "organ",
            "quantity": 1,
            "resource": null,
          },
          Object {
            "gear": null,
            "keyword": "bone",
            "quantity": 1,
            "resource": null,
          },
        ],
        "location": null,
        "not_innovation": null,
        "not_location": null,
      },
    },
  },
  "expansions": Object {
    "core": "core",
  },
  "hunting": null,
  "id": "new",
  "innovations": Object {
    "cooking": "cooking",
  },
  "locations": Object {},
  "notes": "",
  "principles": Object {
    "conviction": null,
    "death": null,
    "newlife": null,
    "society": null,
  },
  "settlement": Object {
    "departing": null,
    "name": "New Settlement",
    "newborn": null,
    "showdown": null,
    "survivalLimit": 2,
    "survivors": Object {},
  },
  "showdown": null,
  "stored_resources": Object {},
}
`;

exports[`with RootStore .selectInnovation adds innovation without bonuses 1`] = `
Object {
  "bonuses": Object {},
  "endeavors": Object {},
  "expansions": Object {
    "core": "core",
  },
  "hunting": null,
  "id": "new",
  "innovations": Object {
    "inner_lantern": "inner_lantern",
  },
  "locations": Object {},
  "notes": "",
  "principles": Object {
    "conviction": null,
    "death": null,
    "newlife": null,
    "society": null,
  },
  "settlement": Object {
    "departing": null,
    "name": "New Settlement",
    "newborn": null,
    "showdown": null,
    "survivalLimit": 1,
    "survivors": Object {},
  },
  "showdown": null,
  "stored_resources": Object {},
}
`;

exports[`with RootStore .selectInnovation removes innovation with bonuses 1`] = `
Object {
  "bonuses": Object {},
  "endeavors": Object {},
  "expansions": Object {
    "core": "core",
  },
  "hunting": null,
  "id": "new",
  "innovations": Object {},
  "locations": Object {},
  "notes": "",
  "principles": Object {
    "conviction": null,
    "death": null,
    "newlife": null,
    "society": null,
  },
  "settlement": Object {
    "departing": null,
    "name": "New Settlement",
    "newborn": null,
    "showdown": null,
    "survivalLimit": 1,
    "survivors": Object {},
  },
  "showdown": null,
  "stored_resources": Object {},
}
`;

exports[`with RootStore .selectInnovation removes innovation without bonuses 1`] = `
Object {
  "bonuses": Object {},
  "endeavors": Object {},
  "expansions": Object {
    "core": "core",
  },
  "hunting": null,
  "id": "new",
  "innovations": Object {},
  "locations": Object {},
  "notes": "",
  "principles": Object {
    "conviction": null,
    "death": null,
    "newlife": null,
    "society": null,
  },
  "settlement": Object {
    "departing": null,
    "name": "New Settlement",
    "newborn": null,
    "showdown": null,
    "survivalLimit": 1,
    "survivors": Object {},
  },
  "showdown": null,
  "stored_resources": Object {},
}
`;

exports[`with RootStore .selectLocation adds location with endeavors 1`] = `
Object {
  "bonuses": Object {},
  "endeavors": Object {
    "harvest_ritual": Object {
      "expansion": "core",
      "id": "harvest_ritual",
      "name": "Harvest Ritual",
      "recipe": Object {
        "innovation": "forbidden dance",
        "items": Array [
          Object {
            "gear": null,
            "keyword": null,
            "quantity": 1,
            "resource": "endeavor",
          },
        ],
        "location": "stone_circle",
        "not_innovation": null,
        "not_location": null,
      },
    },
  },
  "expansions": Object {
    "core": "core",
  },
  "hunting": null,
  "id": "new",
  "innovations": Object {},
  "locations": Object {
    "stone_circle": "stone_circle",
  },
  "notes": "",
  "principles": Object {
    "conviction": null,
    "death": null,
    "newlife": null,
    "society": null,
  },
  "settlement": Object {
    "departing": null,
    "name": "New Settlement",
    "newborn": null,
    "showdown": null,
    "survivalLimit": 1,
    "survivors": Object {},
  },
  "showdown": null,
  "stored_resources": Object {},
}
`;

exports[`with RootStore .selectLocation adds location without endeavors 1`] = `
Object {
  "bonuses": Object {},
  "endeavors": Object {},
  "expansions": Object {
    "core": "core",
  },
  "hunting": null,
  "id": "new",
  "innovations": Object {},
  "locations": Object {
    "catarium": "catarium",
  },
  "notes": "",
  "principles": Object {
    "conviction": null,
    "death": null,
    "newlife": null,
    "society": null,
  },
  "settlement": Object {
    "departing": null,
    "name": "New Settlement",
    "newborn": null,
    "showdown": null,
    "survivalLimit": 1,
    "survivors": Object {},
  },
  "showdown": null,
  "stored_resources": Object {},
}
`;

exports[`with RootStore .selectLocation removes location with endeavors 1`] = `
Object {
  "bonuses": Object {},
  "endeavors": Object {},
  "expansions": Object {
    "core": "core",
  },
  "hunting": null,
  "id": "new",
  "innovations": Object {},
  "locations": Object {},
  "notes": "",
  "principles": Object {
    "conviction": null,
    "death": null,
    "newlife": null,
    "society": null,
  },
  "settlement": Object {
    "departing": null,
    "name": "New Settlement",
    "newborn": null,
    "showdown": null,
    "survivalLimit": 1,
    "survivors": Object {},
  },
  "showdown": null,
  "stored_resources": Object {},
}
`;

exports[`with RootStore .selectLocation removes location without endeavors 1`] = `
Object {
  "bonuses": Object {},
  "endeavors": Object {},
  "expansions": Object {
    "core": "core",
  },
  "hunting": null,
  "id": "new",
  "innovations": Object {},
  "locations": Object {},
  "notes": "",
  "principles": Object {
    "conviction": null,
    "death": null,
    "newlife": null,
    "society": null,
  },
  "settlement": Object {
    "departing": null,
    "name": "New Settlement",
    "newborn": null,
    "showdown": null,
    "survivalLimit": 1,
    "survivors": Object {},
  },
  "showdown": null,
  "stored_resources": Object {},
}
`;

exports[`with RootStore .selectPrinciple adds a principle 1`] = `
Object {
  "bonuses": Object {
    "death": Object {
      "description": Array [
        "Whenever a survivor dies, gain 1 random basic resource",
      ],
      "id": "death",
      "name": "Cannibalize",
    },
  },
  "endeavors": Object {},
  "expansions": Object {
    "core": "core",
  },
  "hunting": null,
  "id": "new",
  "innovations": Object {},
  "locations": Object {},
  "notes": "",
  "principles": Object {
    "conviction": null,
    "death": "cannibalize",
    "newlife": null,
    "society": null,
  },
  "settlement": Object {
    "departing": null,
    "name": "New Settlement",
    "newborn": null,
    "showdown": null,
    "survivalLimit": 2,
    "survivors": Object {},
  },
  "showdown": null,
  "stored_resources": Object {},
}
`;

exports[`with RootStore .selectPrinciple removes a principle with it's bonus 1`] = `
Object {
  "bonuses": Object {},
  "endeavors": Object {},
  "expansions": Object {
    "core": "core",
  },
  "hunting": null,
  "id": "new",
  "innovations": Object {},
  "locations": Object {},
  "notes": "",
  "principles": Object {
    "conviction": null,
    "death": null,
    "newlife": null,
    "society": null,
  },
  "settlement": Object {
    "departing": null,
    "name": "New Settlement",
    "newborn": null,
    "showdown": null,
    "survivalLimit": 1,
    "survivors": Object {},
  },
  "showdown": null,
  "stored_resources": Object {},
}
`;

exports[`with RootStore .showdownMonsterLevel gets the monster level 1`] = `
Object {
  "accuracy": 2,
  "damage": 2,
  "deck": Object {
    "A": 9,
    "B": 10,
    "L": 2,
    "S": Array [
      "Cunning",
      "Merciless",
    ],
  },
  "evasion": 0,
  "huntExtra": null,
  "huntboard": "MMBBMOBMMBBXS",
  "id": "3",
  "luck": 1,
  "movement": 8,
  "name": "3",
  "speed": 2,
  "toughness": 14,
  "type": null,
}
`;
